syntax = "proto2";

package pointpillars;

message Point {
  required float x = 1;
  required float y = 2;
  required float z = 3;
  optional float reflection = 4;
  optional float distance = 5;
}

message Voxel {
  repeated float data = 1 [packed=true];
  required int32 num_voxel = 2;
  required int32 feature_dim = 3;
}

message VoxelCoord {
  repeated int32 data = 1 [packed=true];
  required int32 num_voxel = 2;
  required int32 coord_dim = 3;
}

message VoxelPoints {
  repeated int32 data = 1 [packed=true];
  required int32 num_voxel = 2;
}

message Anchor {
  required float center_x = 1;
  required float center_y = 2;
  required float center_z = 3;
  required float length = 4;
  required float width = 5;
  required float height = 6;
  required float rotation = 7;
  required int32 offset = 8;

  // targeted label id, -1 stands for ignore
  optional int32 target_label = 9 [default = -1];

  // whether it's a postive anchor
  optional bool is_positive = 10 [default = false];
}

message Label {
  required int32 label_id = 1;
  required int32 type = 2;
  required float center_x = 3;
  required float center_y = 4;
  required float center_z = 5;
  required float length = 6;
  required float width = 7;
  required float height = 8;
  required float yaw = 9;
}

message Example {
  required Voxel voxel = 1;  // non-empty continous voxels
  required VoxelCoord voxel_coord = 2;  // x, y coordinats of voxel in sparse input feature map
  required VoxelPoints voxel_points = 3;  // number of non-empty points in each voxel
  repeated Anchor anchor = 4;
  repeated Label label = 5;
  optional string example_id = 6;
  repeated Point point = 7; // raw points is not necessary for training
}
